### DISCUSSION TITLE
FastAPI OpenAI StreamingResponse

### DISCUSSION THREAD

**Discussion by sandeep-srikantam** [fastapi/fastapi#11217] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Generator
from fastapi import APIRouter, Security, Response
from fastapi.responses import StreamingResponse

from example.models.request.chat import ChatRequestModel
from example.models.response.chat import ChatResponseModel
from example.models.common import LinesOfBusiness

from example.core.logger import tracer
from example.config.app import get_settings
from example.services.impl import chat_impl
from example.docs.openapi import openapi_common_responses, openapi_chat_stream_response, openapi_common_request_headers, openapi_security_scheme

chat_stream_router = APIRouter()

def format_as_ndjson(response: Generator[ChatRequestModel, None, None]) -> Generator[str, None,  None]:
    for event in response:
        yield event.model_dump_json() + "\n"

@chat_stream_router.post("/chat/stream", responses={**openapi_common_responses}, description=get_settings().openapi_chat_description)
@tracer
def chat_stream(request: ChatRequestModel, common_headers: openapi_common_request_headers) -> StreamingResponse:

    message_request = chat_impl.run(
        request = request,
        line = line,
        stream = True
    )

    return StreamingResponse(content=format_as_ndjson(message_request), media_type="text/event-stream")
```

### Description

I'm implementing an OpenAI based streaming API and using the StreamingResponse while setting media_type="text/event-stream". This works fine when getting the actual response.

But, I'm not able to sow this up in my Swagger/Redoc UI. It shows the default application/json, but how do I replace it?

I also tried to override the response as following, but this still leaves the default ```application/json``` behind:

```python
openapi_chat_stream_response = {
    status.HTTP_200_OK: {
        "headers": _common_response_headers,
        "content": {
            "text/event-stream": {
                "schema": ChatResponseModel.schema(),
                "example": "" ##Redacted example.
            }
        }
    }
}
```

### Operating System

Linux, Windows

### FastAPI Version

0.110.0

### Pydantic Version

2.6.1

### Python Version

Python 3.11.3

**Comment by vinicius-oa** [comment]
you may need to subclass fastapi's response class:

```python

from fastapi import Response

class TextEventStreamResponse(Response):
    media_type = 'text/event-stream'

@chat_stream_router.post(
    "/chat/stream",
    responses={**openapi_common_responses},
    response_class=TextEventStreamResponse,
    description=get_settings().openapi_chat_description
)
@tracer
def chat_stream(request: ChatRequestModel, common_headers: openapi_common_request_headers) -> StreamingResponse:

    message_request = chat_impl.run(
        request = request,
        line = line,
        stream = True
    )

    return StreamingResponse(content=format_as_ndjson(message_request), media_type="text/event-stream")



```

**Reply by sandeep-srikantam** [reply]
Thank you @vinicius-oa. If we set the response_class, it will change the mime type for all the status codes in that route. Is it possible if can only set the mime type for 200 response and leave other responses at ```application/json```? 

Right now, i pull in my responses from:
```python
from example.models.response.exception import ExceptionResponseModel

openapi_common_responses = {
    status.HTTP_200_OK: {
        "headers": _common_response_headers
    },
    status.HTTP_404_NOT_FOUND: {
        "model": ExceptionResponseModel,
        "headers": _common_response_headers,
    },
    status.HTTP_405_METHOD_NOT_ALLOWED: {
        "model": ExceptionResponseModel,
        "headers": _common_response_headers
    },
    status.HTTP_500_INTERNAL_SERVER_ERROR: {
        "model": ExceptionResponseModel,
        "headers": _common_response_headers
    }
}
```

and then use it in my route as following:
```python
from example.docs.openapi import openapi_common_responses, openapi_chat_stream_response, openapi_common_request_headers, openapi_security_scheme

@chat_stream_router.post("/chat/stream", responses={**openapi_common_responses, **openapi_chat_stream_response}, response_class=TextEventStreamResponse, description=get_settings().openapi_chat_description)
@tracer
def chat_stream(request: ChatRequestModel, common_headers: openapi_common_request_headers, token: str = Security(openapi_security_scheme)) -> ChatResponseModel:

**Reply by vinicius-oa** [reply]
It's not possible, because whenever you set additional responses, fastapi always sets a default response _media_type_ based on the _response_class_ used. In my previous reply, I used _TextEventStreamResponse_, so all additional responses shall also have `text/event-stream`. You can check that at line 363 in fastapi.openapi.utils

![image](https://github.com/tiangolo/fastapi/assets/139140185/600e9052-b139-4d72-bef1-010645989ee7)



However, you may use _TextEventStreamResponse_ as _response_class_ and define _application/json_ in the desired responses, even though in the _/docs_ it will appear two media types for all responses except the _200_, at least will give the illusion of such responses displaying only _application/json_ because the _media_type_ list is ordered alphabetically.  To do that you must in _openapi_common_responses_ variable set an additional key _content_ like the following: `{'content': {'application/json': {'schema': {'$ref': '#/components/schemas/ExceptionResponseModel'}}}`
![image](https://github.com/tiangolo/fastapi/assets/139140185/8661c0cf-2f03-4378-8217-778737a63283)

**Reply by storenth** [reply]
One more problem still here, if you generate an OpenApi client, it suggest the string in the response 
<img width="450" alt="Screenshot 2024-11-12 at 20 57 48" src="https://github.com/user-attachments/assets/abdfab15-9f1e-4b7a-86ef-03ca08102dac">
<img width="954" alt="Screenshot 2024-11-12 at 21 00 11" src="https://github.com/user-attachments/assets/7bb4758b-0a93-49b6-a397-bb3694838bfa">
